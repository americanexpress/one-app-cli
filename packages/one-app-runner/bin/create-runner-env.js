#!/usr/bin/env node

/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

const getPort = require('get-port');
const uuid = require('uuid/v4');
const fs = require('fs-extra');

const createRunnerEnv = async () => {
  const httpPort = await getPort();
  const devCdnPort = await getPort();
  const devProxyServer = await getPort();
  const metricsPort = await getPort();
  const networkName = `one-app-environment-${uuid()}`;

  await fs.writeFile('.env',
    `HTTP_PORT=${httpPort}
HTTP_ONE_APP_DEV_CDN_PORT=${devCdnPort}
NETWORK_NAME=${networkName}
HTTP_ONE_APP_DEV_PROXY_SERVER_PORT=${devProxyServer}
HTTP_METRICS_PORT=${metricsPort}
  `);
};

createRunnerEnv().catch((error) => {
  /* eslint-disable no-console */
  console.log();
  console.error('⚠️   Error creating environment variables: \n', error);
  /* eslint-enable no-console */
});
