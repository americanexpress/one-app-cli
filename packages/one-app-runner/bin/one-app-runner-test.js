#!/usr/bin/env node

/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

require('dotenv').config();
const { spawn } = require('child_process');

const command = `one-app-runner --output-file=one-app-test.log --create-docker-network --docker-network-to-join=${process.env.NETWORK_NAME} --use-host`;
spawn(command, { shell: true, stdio: 'ignore', detached: true }).unref();
