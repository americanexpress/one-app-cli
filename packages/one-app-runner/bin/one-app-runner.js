#!/usr/bin/env node

/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

const path = require('path');
const pkgUp = require('pkg-up');
const yargs = require('yargs');
const startApp = require('../src/startApp');

const createYargsConfig = () => {
  const packageJsonPath = pkgUp.sync();
  // eslint-disable-next-line global-require, import/no-dynamic-require
  const packageJson = require(packageJsonPath);

  const oneAppRunnerConfig = (packageJson['one-amex'] && packageJson['one-amex'].runner) || {};

  if (Object.keys(oneAppRunnerConfig).length > 0) {
    yargs.config(oneAppRunnerConfig);
  }

  // eslint-disable-next-line global-require
  yargs
    .option('module-map-url', {
      type: 'string',
      describe: 'module map for One App to use and fetch modules from',
      coerce: (value) => {
        if (!value) {
          throw new Error('⚠️   --module-map-url option must be given a value but was not given one. Did you mean to pass a value? ⚠️');
        }
        return value;
      },
    })
    .option('root-module-name', {
      type: 'string',
      demandOption: true,
      describe: 'name of the module to serve as an entry point to your application',
      coerce: (value) => {
        if (!value) {
          throw new Error('⚠️   --root-module-name option must be given a value but was not given one. Did you mean to pass a value? ⚠️');
        }
        return value;
      },
    })
    .option('docker-image', {
      type: 'string',
      describe: 'docker image to use for One App',
      demandOption: true,
      coerce: (value) => {
        if (!value) {
          throw new Error('⚠️   --docker-image option must be given a value but was not given one. Did you mean to pass a value? ⚠️');
        }
        return value;
      },
    })
    .option('parrot-middleware', {
      describe: 'path to parrot dev middleware file for One App to use for Parrot mocking',
      type: 'string',
      coerce: (value) => {
        if (!value) {
          return value;
        }
        return path.resolve(process.cwd(), value);
      },
    })
    .option('modules', {
      type: 'array',
      // eslint-disable-next-line global-require
      demandOption: !require('yargs').argv.moduleMapUrl,
      describe: 'path to local module to serve to One App',
      coerce: (modules) => {
        if (!modules.length > 0) {
          throw new Error('⚠️   --modules option must be given a value but was not given one. Did you mean to pass a value? ⚠️');
        }
        const processedModules = modules.map((module) => path.resolve(process.cwd(), module));
        return processedModules;
      },
    })
    .option('dev-endpoints', {
      describe: 'path to dev endpoints file for One App to use for its One App Dev Proxy set up',
      type: 'string',
      coerce: (value) => {
        if (!value) {
          throw new Error('⚠️   --dev-endpoints option must be given a value but was not given one. Did you mean to pass a value? ⚠️');
        }
        return path.resolve(process.cwd(), value);
      },
    })
    .option('envVars', {
      describe: 'Environment variables to be applied to the One App instance',
      coerce: (value) => {
        if (typeof value === 'object') {
          return value;
        }
        try {
          return JSON.parse(value);
        } catch (err) {
          throw new Error('⚠️   --envVars option failed to parse a value. Reference the Readme for valid configuration examples. ⚠️');
        }
      },
    })
    .option('output-file', {
      type: 'string',
      describe: 'File to redirect all stdout and stderr from One App Container to',
      coerce: (value) => path.resolve(process.cwd(), value),
    })
    .option('create-docker-network', {
      default: false,
      type: 'boolean',
      describe: 'Creates a new docker network',
    })
    .option('docker-network-to-join', {
      type: 'string',
      describe: 'Connect One App container to a docker network',
    })
    .option('use-host', {
      describe: 'Use req.headers.host instead of localhost for one-app-dev-cdn',
      default: false,
      type: 'boolean',
    })
    .option('offline', {
      describe: 'skip docker pull when the docker registry is not available / offline',
      default: false,
      type: 'boolean',
    })
    .option('--container-name', {
      describe: 'Assign a container name with the --name option',
      type: 'string',
    })
    .implies({
      'parrot-middleware': 'modules',
      'dev-endpoints': 'modules',
    })
    .strict()
    .help();

  return yargs.argv;
};

const argv = createYargsConfig();

if (!argv.envVars) {
  argv.envVars = {};
}

try {
  startApp({
    moduleMapUrl: argv.moduleMapUrl,
    rootModuleName: argv.rootModuleName,
    modulesToServe: argv.modules,
    parrotMiddlewareFile: argv.parrotMiddleware,
    devEndpointsFile: argv.devEndpoints,
    appDockerImage: argv.dockerImage,
    envVars: argv.envVars,
    outputFile: argv.outputFile,
    createDockerNetwork: argv.createDockerNetwork,
    dockerNetworkToJoin: argv.dockerNetworkToJoin,
    useHost: argv.useHost,
    offline: argv.offline,
    containerName: argv.containerName,
  });
} catch (error) {
  /* eslint-disable no-console */
  console.log();
  console.error('⚠️   Error starting up your One App environment: \n', error.message);
  /* eslint-enable no-console */
}
