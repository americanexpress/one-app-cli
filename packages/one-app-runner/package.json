{
  "name": "@americanexpress/one-app-runner",
  "version": "6.12.0",
  "description": "CLI for running One App locally",
  "license": "Apache-2.0",
  "contributors": [
    "Andres Escobar <Andres.Escobar@aexp.com> (https://github.com/anescobar1991)",
    "James Singleton <James.Singleton1@aexp.com> (https://github.com/JamesSingleton)",
    "Jimmy King  <Jimmy.King@aexp.com> (https://github.com/10xLaCroixDrinker)",
    "Jonathan Adshead <Jonathan.Adshead@aexp.com> (https://github.com/JAdshead)",
    "Michael Tobia <Michael.M.Tobia@aexp.com> (https://github.com/Francois-Esquire)",
    "Michael Tomcal <Michael.A.Tomcal@aexp.com> (https://github.com/mtomcal)",
    "Stephanie Coates  <Stephanie.Coates1@aexp.com> (https://github.com/stephaniecoates)",
    "Nelly Kiboi <Nelly.J.Kiboi@aexp.com> (https://github.com/nellyk)"
  ],
  "bin": {
    "one-app-runner": "./bin/one-app-runner.js",
    "create-runner-env": "./bin/create-runner-env.js",
    "one-app-runner-test": "bin/one-app-runner-test.js"
  },
  "files": [
    "src",
    "bin"
  ],
  "repository": {
    "url": "https://github.com/americanexpress/one-app-cli.git",
    "directory": "packages/one-app-runner"
  },
  "keywords": [
    "dx",
    "one-app",
    "docker"
  ],
  "dependencies": {
    "dockerode": "^3.2.0",
    "dotenv": "^8.2.0",
    "fs-extra": "^9.0.1",
    "get-port": "^5.1.1",
    "pkg-up": "^3.1.0",
    "uuid": "^8.3.2",
    "yargs": "^16.2.0"
  },
  "devDependencies": {
    "mock-spawn": "^0.2.6"
  },
  "publishConfig": {
    "access": "public"
  }
}
