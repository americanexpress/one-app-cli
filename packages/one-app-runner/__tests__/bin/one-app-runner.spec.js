/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

// explicitly requiring within each test needed in order to have independent mocks
/* eslint-disable global-require */
const path = require('path');

const originalProcessExit = process.exit;
const originalProcessArgv = process.argv;

beforeAll(() => {
  process.exit = jest.fn();
  jest.spyOn(console, 'error');
  console.error.mockImplementation();
});

beforeEach(() => {
  jest
    .resetModules()
    .clearAllMocks();

  jest.unmock('../../../../package.json');

  process.argv = originalProcessArgv;
});

afterAll(() => {
  process.argv = originalProcessArgv;
  process.exit = originalProcessExit;
});

test('command errors out if --module-map-url option is not given a value', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error');

  process.argv = ['', '', '--module-map-url'];
  jest.mock('../../src/startApp', () => jest.fn());
  require('../../bin/one-app-runner');
  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();
});

test('command errors out if --module-map-url option is not given', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error');

  process.argv = ['', '', '--root-module-name', 'frank-lloyd-root', '--docker-image', 'one-app:5.0.0'];
  jest.mock('../../src/startApp', () => jest.fn());
  require('../../bin/one-app-runner');
  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();
});

test('command errors out if --root-module-name option is not given a value', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error');

  process.argv = ['', '', '--root-module-name'];
  jest.mock('../../src/startApp', () => jest.fn());
  require('../../bin/one-app-runner');
  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();
});

test('command errors out if --root-module-name option is not given', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error');

  process.argv = ['', '', '--module-map-url', 'https://example.com/module-map.json', '--docker-image', 'one-app:5.0.0'];
  jest.mock('../../src/startApp', () => jest.fn());
  require('../../bin/one-app-runner');
  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();
});

test('command errors out if --docker-image option is not given a value', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error');

  process.argv = ['', '', '--docker-image'];
  jest.mock('../../src/startApp', () => jest.fn());
  require('../../bin/one-app-runner');
  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();
});

test('command errors out if --docker-image option is not given', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error');

  process.argv = ['', '', '--module-map-url', 'https://example.com/module-map.json', '--root-module-name', 'frank-lloyd-root'];
  jest.mock('../../src/startApp', () => jest.fn());
  require('../../bin/one-app-runner');
  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();
});

test('--parrot-middleware, --modules, --output-file, and --dev-endpoints values are coerced into absolute paths', () => {
  const modulesPath = '../fake/path/to/fake-module';
  const parrotMiddlewarePath = '../fake/path/to/fake-module/dev.middleware.js';
  const outputFile = '../fake/path/to/fake-module/one-app.log';
  const devEndpointsFilePath = '../fake/path/to/fake-module/dev.endpoints.js';

  process.argv = [
    '',
    '',
    '--modules',
    modulesPath,
    '--parrot-middleware',
    parrotMiddlewarePath,
    '--output-file',
    outputFile,
    '--dev-endpoints',
    devEndpointsFilePath,
  ];
  jest.mock('../../src/startApp', () => jest.fn());
  const startApp = require('../../src/startApp');

  require('../../bin/one-app-runner');

  expect(
    path.isAbsolute(startApp.mock.calls[0][0].parrotMiddlewareFile)
  ).toBeTruthy();
  expect(
    path.isAbsolute(startApp.mock.calls[0][0].devEndpointsFile)
  ).toBeTruthy();
  expect(
    path.isAbsolute(startApp.mock.calls[0][0].modulesToServe[0])
  ).toBeTruthy();
});

test('command errors out if --dev-endpoints is not given a value', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error');

  process.argv = ['', '', '--dev-endpoints'];
  jest.mock('../../src/startApp', () => jest.fn(() => Promise.resolve()));
  require('../../bin/one-app-runner');
  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();
});

test('reads modules from package.json', async () => {
  const mockModulePath = path.resolve('/fake/path/to/fake-module');
  jest.mock('../../../../package.json', () => ({
    'one-amex': {
      runner: {
        modules: [
          mockModulePath,
        ],
      },
    },
  }));

  jest.mock('../../src/startApp', () => jest.fn());
  const startApp = require('../../src/startApp');
  await require('../../bin/one-app-runner');
  expect(startApp.mock.calls[0][0].modulesToServe).toEqual([mockModulePath]);
});

test('command errors out if --modules option is not given any values', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error');

  process.argv = ['', '', '--modules'];
  jest.mock('../../src/startApp', () => jest.fn(() => Promise.resolve()));
  require('../../bin/one-app-runner');
  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();
});

test('--parrot-middleware, --module values, --output-file, and --dev-endpoints are left alone if already absolute paths', () => {
  const modulesPath = path.resolve('/fake/path/to/fake-module');
  const parrotMiddlewarePath = path.resolve('/fake/path/to/fake-module/dev.middleware.js');
  const outputFile = path.resolve('/fake/path/to/fake-module/one-app.log');
  const devEndpointsFilePath = path.resolve('/fake/path/to/fake-module/dev.endpoints.js');

  process.argv = [
    '',
    '',
    '--modules',
    modulesPath,
    '--parrot-middleware',
    parrotMiddlewarePath,
    '--output-file',
    outputFile,
    '--dev-endpoints',
    devEndpointsFilePath,
  ];
  jest.mock('../../src/startApp', () => jest.fn());
  const startApp = require('../../src/startApp');
  require('../../bin/one-app-runner');

  expect(
    startApp.mock.calls[0][0].modulesToServe[0]
  ).toBe(modulesPath);
  expect(
    startApp.mock.calls[0][0].parrotMiddlewareFile
  ).toBe(parrotMiddlewarePath);
  expect(
    startApp.mock.calls[0][0].devEndpointsFile
  ).toBe(devEndpointsFilePath);
});

test('all options are used if specified', () => {
  const dockerImage = 'one-app:5.0.0';
  const devEndpointsFile = path.resolve('/fake/path/to/fake-module/dev.endpoints.js');
  const envVars = 'MY_ENV_VAR=value';
  const moduleMapUrl = 'httops://example.com/module-map.json';
  const modulesPath = path.resolve('/fake/path/to/fake-module');
  const outputFile = path.resolve('/fake/path/to/fake-module/one-app.log');
  const parrotMiddlewarePath = path.resolve('/fake/path/to/fake-module/dev.middleware.js');
  const rootModuleName = 'frank-lloyd-root';
  const dockerNetworkToJoin = 'one-test-environment-1234';

  process.argv = [
    '',
    '',
    '--docker-image',
    dockerImage,
    '--dev-endpoints',
    devEndpointsFile,
    '--env-vars',
    envVars,
    '--module-map-url',
    moduleMapUrl,
    '--modules',
    modulesPath,
    '--output-file',
    outputFile,
    '--parrot-middleware',
    parrotMiddlewarePath,
    '--root-module-name',
    rootModuleName,
    '--docker-network-to-join',
    dockerNetworkToJoin,
    '--use-host',
    '--offline',
  ];

  jest.mock('../../src/startApp', () => jest.fn());
  const startApp = require('../../src/startApp');
  require('../../bin/one-app-runner');

  expect(startApp.mock.calls).toMatchSnapshot();
});

test('command errors out if an unknown option is given', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error');

  process.argv = ['', '', '--not-a-valid-option', '--root-module-name', 'frank-lloyd-root', '--module-map-url', 'https://example.com/module-map.json', '--docker-image', 'one-app:5.0.0'];
  jest.mock('../../src/startApp', () => jest.fn());
  require('../../bin/one-app-runner');
  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();
});

test('--modules option is required if --module-map-url option is not given', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error');

  process.argv = ['', '', '--root-module-name', 'frank-lloyd-root', '--docker-image', 'one-app:5.0.0'];
  jest.mock('../../src/startApp', () => jest.fn());
  require('../../bin/one-app-runner');
  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();
});

test('--modules option is required if --parrot-middleware option is given', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error');

  process.argv = ['', '', '--parrot-middleware', '../path/to/dev.middleware.js', '--root-module-name', 'frank-lloyd-root', '--module-map-url', 'https://example.com/module-map.json', '--docker-image', 'one-app:5.0.0'];
  jest.mock('../../src/startApp', () => jest.fn());
  require('../../bin/one-app-runner');
  expect(consoleErrorSpy.mock.calls[2][0]).toContain('parrot-middleware -> modules');
});

test('--modules option is required if --dev-endpoints option is given', () => {
  const consoleErrorSpy = jest.spyOn(console, 'error');

  process.argv = ['', '', '--dev-endpoints', '../path/to/dev.endpoints.js', '--root-module-name', 'frank-lloyd-root', '--module-map-url', 'https://example.com/module-map.json', '--docker-image', 'one-app:5.0.0'];
  jest.mock('../../src/startApp', () => jest.fn());
  require('../../bin/one-app-runner');
  expect(consoleErrorSpy.mock.calls[2][0]).toContain('dev-endpoints -> modules');
});

test('an `envVars` key is supported within the config entry', () => {
  jest.mock('../../../../package.json', () => ({
    'one-amex': {
      runner: {
        envVars: {
          MY_COOL_CONFIG_VARIABLE: 'is-the-best',
        },
      },
    },
  }));

  jest.mock('../../src/startApp', () => jest.fn());
  const startApp = require('../../src/startApp');
  require('../../bin/one-app-runner');

  expect(startApp.mock.calls[0][0].envVars)
    .toHaveProperty('MY_COOL_CONFIG_VARIABLE', 'is-the-best');
});

test('an `envVars` arg is supported', () => {
  process.argv = [
    '',
    '',
    '--envVars',
    '{ "MY_VARIABLE": "my-variable", "HELLO": "WORLD" }',
  ];

  jest.mock('../../src/startApp', () => jest.fn());
  const startApp = require('../../src/startApp');

  require('../../bin/one-app-runner');
  expect(startApp.mock.calls[0][0].envVars).toMatchSnapshot();
});
