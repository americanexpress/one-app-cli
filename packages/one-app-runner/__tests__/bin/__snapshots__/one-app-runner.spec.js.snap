// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`--modules option is required if --dev-endpoints option is given 1`] = `
Array [
  Array [
    "Options:
  --version                 Show version number                        [boolean]
  --module-map-url          module map for One App to use and fetch modules from
                                                                        [string]
  --root-module-name        name of the module to serve as an entry point to
                            your application                 [string] [required]
  --docker-image            docker image to use for One App  [string] [required]
  --parrot-middleware       path to parrot dev middleware file for One App to
                            use for Parrot mocking                      [string]
  --modules                 path to local module to serve to One App     [array]
  --dev-endpoints           path to dev endpoints file for One App to use for
                            its One App Dev Proxy set up                [string]
  --envVars                 Environment variables to be applied to the One App
                            instance
  --output-file             File to redirect all stdout and stderr from One App
                            Container to                                [string]
  --create-docker-network   Creates a new docker network
                                                      [boolean] [default: false]
  --docker-network-to-join  Connect One App container to a docker network
                                                                        [string]
  --use-host                Use req.headers.host instead of localhost for
                            one-app-dev-cdn           [boolean] [default: false]
  --offline                 skip docker pull when the docker registry is not
                            available / offline       [boolean] [default: false]
  ----container-name        Assign a container name with the --name option
                                                                        [string]
  --help                    Show help                                  [boolean]",
  ],
  Array [],
  Array [
    "Missing dependent arguments:
 dev-endpoints -> modules",
  ],
]
`;

exports[`--modules option is required if --module-map-url option is not given 1`] = `
Array [
  Array [
    "Options:
  --version                 Show version number                        [boolean]
  --module-map-url          module map for One App to use and fetch modules from
                                                                        [string]
  --root-module-name        name of the module to serve as an entry point to
                            your application                 [string] [required]
  --docker-image            docker image to use for One App  [string] [required]
  --parrot-middleware       path to parrot dev middleware file for One App to
                            use for Parrot mocking                      [string]
  --modules                 path to local module to serve to One App
                                                              [array] [required]
  --dev-endpoints           path to dev endpoints file for One App to use for
                            its One App Dev Proxy set up                [string]
  --envVars                 Environment variables to be applied to the One App
                            instance
  --output-file             File to redirect all stdout and stderr from One App
                            Container to                                [string]
  --create-docker-network   Creates a new docker network
                                                      [boolean] [default: false]
  --docker-network-to-join  Connect One App container to a docker network
                                                                        [string]
  --use-host                Use req.headers.host instead of localhost for
                            one-app-dev-cdn           [boolean] [default: false]
  --offline                 skip docker pull when the docker registry is not
                            available / offline       [boolean] [default: false]
  ----container-name        Assign a container name with the --name option
                                                                        [string]
  --help                    Show help                                  [boolean]",
  ],
  Array [],
  Array [
    "Missing required argument: modules",
  ],
]
`;

exports[`--modules option is required if --parrot-middleware option is given 1`] = `
Array [
  Array [
    "Options:
  --version                 Show version number                        [boolean]
  --module-map-url          module map for One App to use and fetch modules from
                                                                        [string]
  --root-module-name        name of the module to serve as an entry point to
                            your application                 [string] [required]
  --docker-image            docker image to use for One App  [string] [required]
  --parrot-middleware       path to parrot dev middleware file for One App to
                            use for Parrot mocking                      [string]
  --modules                 path to local module to serve to One App     [array]
  --dev-endpoints           path to dev endpoints file for One App to use for
                            its One App Dev Proxy set up                [string]
  --envVars                 Environment variables to be applied to the One App
                            instance
  --output-file             File to redirect all stdout and stderr from One App
                            Container to                                [string]
  --create-docker-network   Creates a new docker network
                                                      [boolean] [default: false]
  --docker-network-to-join  Connect One App container to a docker network
                                                                        [string]
  --use-host                Use req.headers.host instead of localhost for
                            one-app-dev-cdn           [boolean] [default: false]
  --offline                 skip docker pull when the docker registry is not
                            available / offline       [boolean] [default: false]
  ----container-name        Assign a container name with the --name option
                                                                        [string]
  --help                    Show help                                  [boolean]",
  ],
  Array [],
  Array [
    "Missing dependent arguments:
 parrot-middleware -> modules",
  ],
]
`;

exports[`all options are used if specified 1`] = `
Array [
  Array [
    Object {
      "appDockerImage": "one-app:5.0.0",
      "containerName": undefined,
      "createDockerNetwork": false,
      "devEndpointsFile": "/fake/path/to/fake-module/dev.endpoints.js",
      "dockerNetworkToJoin": "one-test-environment-1234",
      "envVars": "MY_ENV_VAR=value",
      "moduleMapUrl": "httops://example.com/module-map.json",
      "modulesToServe": Array [
        "/fake/path/to/fake-module",
      ],
      "offline": true,
      "outputFile": "/fake/path/to/fake-module/one-app.log",
      "parrotMiddlewareFile": "/fake/path/to/fake-module/dev.middleware.js",
      "rootModuleName": "frank-lloyd-root",
      "useHost": true,
    },
  ],
]
`;

exports[`an \`envVars\` arg is supported 1`] = `
Object {
  "HELLO": "WORLD",
  "MY_VARIABLE": "my-variable",
}
`;

exports[`command errors out if --dev-endpoints is not given a value 1`] = `
Array [
  Array [
    "Options:
  --help               Show help                                       [boolean]
  --version            Show version number                             [boolean]
  --module-map-url     module map for One App to use and fetch modules from
                                                                        [string]
  --root-module-name   name of the module to serve as an entry point to your
                       application                           [string] [required]
  --docker-image       docker image to use for One App       [string] [required]
  --parrot-middleware  path to parrot dev middleware file for One App to use for
                       Parrot mocking                                   [string]",
  ],
  Array [],
  Array [
    "Missing required arguments: root-module-name, docker-image",
  ],
]
`;

exports[`command errors out if --docker-image option is not given 1`] = `
Array [
  Array [
    "Options:
  --help               Show help                                       [boolean]
  --version            Show version number                             [boolean]
  --module-map-url     module map for One App to use and fetch modules from
                                                                        [string]
  --root-module-name   name of the module to serve as an entry point to your
                       application                           [string] [required]
  --docker-image       docker image to use for One App       [string] [required]
  --parrot-middleware  path to parrot dev middleware file for One App to use for
                       Parrot mocking                                   [string]",
  ],
  Array [],
  Array [
    "Missing required argument: docker-image",
  ],
]
`;

exports[`command errors out if --docker-image option is not given a value 1`] = `
Array [
  Array [
    "Options:
  --help               Show help                                       [boolean]
  --version            Show version number                             [boolean]
  --module-map-url     module map for One App to use and fetch modules from
                                                                        [string]
  --root-module-name   name of the module to serve as an entry point to your
                       application                           [string] [required]
  --docker-image       docker image to use for One App       [string] [required]
  --parrot-middleware  path to parrot dev middleware file for One App to use for
                       Parrot mocking                                   [string]",
  ],
  Array [],
  Array [
    "⚠️   --docker-image option must be given a value but was not given one. Did you mean to pass a value? ⚠️",
  ],
]
`;

exports[`command errors out if --module-map-url option is not given 1`] = `
Array [
  Array [
    "Options:
  --version                 Show version number                        [boolean]
  --module-map-url          module map for One App to use and fetch modules from
                                                                        [string]
  --root-module-name        name of the module to serve as an entry point to
                            your application                 [string] [required]
  --docker-image            docker image to use for One App  [string] [required]
  --parrot-middleware       path to parrot dev middleware file for One App to
                            use for Parrot mocking                      [string]
  --modules                 path to local module to serve to One App
                                                              [array] [required]
  --dev-endpoints           path to dev endpoints file for One App to use for
                            its One App Dev Proxy set up                [string]
  --envVars                 Environment variables to be applied to the One App
                            instance
  --output-file             File to redirect all stdout and stderr from One App
                            Container to                                [string]
  --create-docker-network   Creates a new docker network
                                                      [boolean] [default: false]
  --docker-network-to-join  Connect One App container to a docker network
                                                                        [string]
  --use-host                Use req.headers.host instead of localhost for
                            one-app-dev-cdn           [boolean] [default: false]
  --offline                 skip docker pull when the docker registry is not
                            available / offline       [boolean] [default: false]
  ----container-name        Assign a container name with the --name option
                                                                        [string]
  --help                    Show help                                  [boolean]",
  ],
  Array [],
  Array [
    "Missing required argument: modules",
  ],
]
`;

exports[`command errors out if --module-map-url option is not given a value 1`] = `
Array [
  Array [
    "Options:
  --help               Show help                                       [boolean]
  --version            Show version number                             [boolean]
  --module-map-url     module map for One App to use and fetch modules from
                                                                        [string]
  --root-module-name   name of the module to serve as an entry point to your
                       application                           [string] [required]
  --docker-image       docker image to use for One App       [string] [required]
  --parrot-middleware  path to parrot dev middleware file for One App to use for
                       Parrot mocking                                   [string]",
  ],
  Array [],
  Array [
    "⚠️   --module-map-url option must be given a value but was not given one. Did you mean to pass a value? ⚠️",
  ],
]
`;

exports[`command errors out if --modules option is not given any values 1`] = `
Array [
  Array [
    "Options:
  --help               Show help                                       [boolean]
  --version            Show version number                             [boolean]
  --module-map-url     module map for One App to use and fetch modules from
                                                                        [string]
  --root-module-name   name of the module to serve as an entry point to your
                       application                           [string] [required]
  --docker-image       docker image to use for One App       [string] [required]
  --parrot-middleware  path to parrot dev middleware file for One App to use for
                       Parrot mocking                                   [string]",
  ],
  Array [],
  Array [
    "Missing required arguments: root-module-name, docker-image",
  ],
]
`;

exports[`command errors out if --root-module-name option is not given 1`] = `
Array [
  Array [
    "Options:
  --help               Show help                                       [boolean]
  --version            Show version number                             [boolean]
  --module-map-url     module map for One App to use and fetch modules from
                                                                        [string]
  --root-module-name   name of the module to serve as an entry point to your
                       application                           [string] [required]
  --docker-image       docker image to use for One App       [string] [required]
  --parrot-middleware  path to parrot dev middleware file for One App to use for
                       Parrot mocking                                   [string]",
  ],
  Array [],
  Array [
    "Missing required argument: root-module-name",
  ],
]
`;

exports[`command errors out if --root-module-name option is not given a value 1`] = `
Array [
  Array [
    "Options:
  --help               Show help                                       [boolean]
  --version            Show version number                             [boolean]
  --module-map-url     module map for One App to use and fetch modules from
                                                                        [string]
  --root-module-name   name of the module to serve as an entry point to your
                       application                           [string] [required]
  --docker-image       docker image to use for One App       [string] [required]
  --parrot-middleware  path to parrot dev middleware file for One App to use for
                       Parrot mocking                                   [string]",
  ],
  Array [],
  Array [
    "⚠️   --root-module-name option must be given a value but was not given one. Did you mean to pass a value? ⚠️",
  ],
]
`;

exports[`command errors out if an unknown option is given 1`] = `
Array [
  Array [
    "Options:
  --version                 Show version number                        [boolean]
  --module-map-url          module map for One App to use and fetch modules from
                                                                        [string]
  --root-module-name        name of the module to serve as an entry point to
                            your application                 [string] [required]
  --docker-image            docker image to use for One App  [string] [required]
  --parrot-middleware       path to parrot dev middleware file for One App to
                            use for Parrot mocking                      [string]
  --modules                 path to local module to serve to One App     [array]
  --dev-endpoints           path to dev endpoints file for One App to use for
                            its One App Dev Proxy set up                [string]
  --envVars                 Environment variables to be applied to the One App
                            instance
  --output-file             File to redirect all stdout and stderr from One App
                            Container to                                [string]
  --create-docker-network   Creates a new docker network
                                                      [boolean] [default: false]
  --docker-network-to-join  Connect One App container to a docker network
                                                                        [string]
  --use-host                Use req.headers.host instead of localhost for
                            one-app-dev-cdn           [boolean] [default: false]
  --offline                 skip docker pull when the docker registry is not
                            available / offline       [boolean] [default: false]
  ----container-name        Assign a container name with the --name option
                                                                        [string]
  --help                    Show help                                  [boolean]",
  ],
  Array [],
  Array [
    "Unknown arguments: not-a-valid-option, notAValidOption",
  ],
]
`;
